docs/api_reference/src/quantities.tex

```diff
@@ -36,107 +39,2549 @@
+  !@\exposconceptnc{NestedQuantityKindSpecOf}@<T{}, QS> &&
+  (@\exposconceptnc{QuantityKindSpec}@<T> || !@\exposconceptnc{NestedQuantityKindSpecOf}@<QS, T{}>);
+\end{itemdecl}
+
+\rSec2[qty.spec.types]{Types}
+
+\begin{codeblock}
+namespace mp_units {
+
+template<@\libconcept{QuantitySpec}@ QS, @\libconcept{Unit}@ U>
+  requires(!@\libconcept{AssociatedUnit}@<U>) || @\libconcept{UnitOf}@<U, QS{}>
+consteval @\libconcept{Reference}@ auto @\exposidnc{make-reference}@(QS, U u)
+{
+  if constexpr (@\exposconceptnc{QuantityKindSpec}@<QS>)
+    return u;
+  else
+    return reference<QS, U>{};
+}
+
+consteval quantity_character @\exposidnc{common-quantity-character}@(
+  std::initializer_list<quantity_character> args)
+{
+  return max(args);
+}
+
+template<typename... Qs1, typename... Qs2>
+consteval quantity_character @\exposidnc{derived-quantity-character}@(@\exposidnc{type-list}@<Qs1...>, @\exposidnc{type-list}@<Qs2...>)
+{
+  constexpr quantity_character num =
+    @\exposidnc{common-quantity-character}@({quantity_character::scalar, [:@\exposidnc{expr-type}@(^Qs1):]::character...});
+  constexpr quantity_character den =
+    @\exposidnc{common-quantity-character}@({quantity_character::scalar, [:@\exposidnc{expr-type}@(^Qs2):]::character...});
+  if constexpr (num == den)
+    return quantity_character::scalar;
+  else
+    return @\exposidnc{common-quantity-character}@({num, den});
+}
+
+consteval quantity_character @\exposidnc{quantity-character-init}@(quantity_character ch,
+                                                     std::initializer_list<std::meta::info> args)
+{
+  auto it =
+    std::ranges::find_if(args, [](auto arg) { return type_of(arg) == ^quantity_character; });
+  if (it != args.end()) return extract<quantity_character>(*it);
+  return ch;
+}
+
+template<@\defexposconceptnc{NamedQuantitySpec}@ Q>
+  requires requires { Q::dimension; }
+using @\exposidnc{to-dimension}@ = [:type_remove_const(type_of(^Q::dimension)):];
+
+struct @\exposidnc{quantity-spec-interface-base}@ {
+  friend consteval @\libconcept{QuantitySpec}@ auto operator*(@\libconcept{QuantitySpec}@ auto lhs, @\libconcept{QuantitySpec}@ auto rhs)
+  {
+    return @\exposidnc{clone-kind-of}@<lhs, rhs>(@\exposidnc{expr-multiply}@<derived_quantity_spec, struct dimensionless>(
+      @\exposidnc{remove-kind}@(lhs), @\exposidnc{remove-kind}@(rhs)));
+  }
+
+  friend consteval @\libconcept{QuantitySpec}@ auto operator/(@\libconcept{QuantitySpec}@ auto lhs, @\libconcept{QuantitySpec}@ auto rhs)
+  {
+    return @\exposidnc{clone-kind-of}@<lhs, rhs>(@\exposidnc{expr-divide}@<derived_quantity_spec, struct dimensionless>(
+      @\exposidnc{remove-kind}@(lhs), @\exposidnc{remove-kind}@(rhs)));
+  }
+
+  friend consteval bool operator==(@\libconcept{QuantitySpec}@ auto lhs, @\libconcept{QuantitySpec}@ auto rhs)
+  {
+    return ^decltype(lhs) == ^decltype(rhs);
+  }
+};
+
+struct @\exposidnc{quantity-spec-interface}@ : @\exposidnc{quantity-spec-interface-base}@ {
```

Comment:
`quantity-spec-interface-base` is not needed here. I did it only for older compilers where `quantity-spec-interface` is a template, which would prevent me from implementing `QuantitySpec`.

Here,  `quantity-spec-interface` is not a template, so it may contain all of the implementation, and the `QuantitySpec` concept may use it in its definition.